numpy/lib/shape_base.py

```diff
@@ -376,7 +377,15 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
         ind0 = next(inds)
     except StopIteration:
         raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
-    res = asanyarray(func1d(inarr_view[ind0], *args, **kwargs))
+
+    res = func1d(inarr_view[ind0], *args, **kwargs)
+    if res is None:
+        # no return value: assume arr is being modified in-place
```

Comment:
Not a safe assumption, `func1d` could return a mixture of `None`s and other objects

Reply:
I'm not entirely sure I follow you. You mean in that checking only the first call to func1d may not be representative of the return value for all locations?

If you think it is preferable, I can revert to a keyword-based approach as implemented in 7d910da

Reply:
It's not clear to me that this function makes sense at all. Why are we working a glorified loop with a callback and not an iterator?